src/useTimer.ts

```diff
@@ -19,92 +19,58 @@ export interface IValues {
-    ...initialConfig,
-    ...config,
-  };
-  const intervalRef = useRef<NodeJS.Timeout | null>(null);
-  const pausedTimeRef = useRef<number | null>(null);
-  const [shouldResetTime, setShouldResetTime] = useState(false);
+export const useTimer = ({
+  initialTime = 0,
+  interval = 1000,
+  step = 1,
+  timerType = 'INCREMENTAL',
+  endTime,
+  onTimeOver,
+}: Partial<IConfig> = {}): IValues => {
   const [time, setTime] = useState(initialTime);
   const [isRunning, setIsRunning] = useState(false);
+  const [isCountFinished, setIsCountFinished] = useState(false);
 
-  const cancelInterval = () => {
-    if (intervalRef.current) {
-      clearInterval(intervalRef.current);
-      intervalRef.current = null;
-      setIsRunning(false);
-    }
-  };
-
-  const createInterval = () => {
-    setIsRunning(true);
-
-    intervalRef.current = setInterval(() => {
-      setTime(previousTime => {
-        const newTime =
-          timerType === 'INCREMENTAL'
-            ? previousTime + step
-            : previousTime - step;
-
-        if (endTime !== null && newTime === endTime) {
-          stopTimerWhenTimeIsOver();
-        }
-
-        return newTime;
-      });
-    }, interval);
-  };
-
-  const pause = () => {
-    pausedTimeRef.current = time;
-
-    cancelInterval();
-  };
-
-  const reset = () => {
-    pausedTimeRef.current = null;
-
-    cancelInterval();
-    resetTime();
-  };
-
-  const resetTime = () => {
+  const reset = useCallback(() => {
+    setIsRunning(false);
+    setIsCountFinished(false);
     setTime(initialTime);
-  };
-
-  const start = () => {
-    if (intervalRef.current) {
-      return;
-    }
+  }, [initialTime]);
 
-    if (shouldResetTime) {
-      resetTime();
-      setShouldResetTime(false);
+  const start = useCallback(() => {
+    if (isCountFinished) {
+      reset();
     }
+    setIsRunning(true);
+  }, [reset, isCountFinished]);
 
-    createInterval();
-  };
-
-  const stopTimerWhenTimeIsOver = () => {
-    cancelInterval();
-    setShouldResetTime(true);
+  const pause = useCallback(() => {
+    setIsRunning(false);
+  }, []);
 
-    if (typeof onTimeOver === 'function') {
-      onTimeOver();
+  useEffect(() => {
+    if (isRunning && time === endTime) {
+      setIsRunning(false);
+      setIsCountFinished(true);
+      if (onTimeOver) onTimeOver();
     }
-  };
-
-  useEffect(() => cancelInterval, []);
+  }, [endTime, onTimeOver, time, reset, isRunning]);
+
+  useEffect(() => {
+    let intervalId: NodeJS.Timeout | null = null;
+    if (isRunning) {
+      intervalId = setInterval(() => {
+        setTime(time =>
+          timerType === 'DECREMENTAL' ? time - step : time + step
+        );
+      }, interval);
+    } else if (!isRunning) {
+      if (intervalId) clearInterval(intervalId);
+    }
+    return () => {
+      if (intervalId) clearInterval(intervalId);
+    };
+  }, [isRunning, time, step, timerType, interval]);
```

Comment:
Inside this dependencies array, you don't need **time** variable, as you get fresh value from **setTime** function.

```suggestion
  }, [isRunning, step, timerType, interval]);
```